homeassistant/components/devolo_home_control/config_flow.py

```diff
@@ -29,22 +35,30 @@ def __init__(self):
     async def async_step_user(self, user_input=None):
         """Handle a flow initiated by the user."""
         if self.show_advanced_options:
-            self.data_schema = {
-                vol.Required(CONF_USERNAME): str,
-                vol.Required(CONF_PASSWORD): str,
-                vol.Required(CONF_MYDEVOLO, default=DEFAULT_MYDEVOLO): str,
-            }
+            self.data_schema[
+                vol.Required(CONF_MYDEVOLO, default=DEFAULT_MYDEVOLO)
+            ] = str
         if user_input is None:
             return self._show_form(user_input)
-        user = user_input[CONF_USERNAME]
-        password = user_input[CONF_PASSWORD]
-        mydevolo = Mydevolo()
-        mydevolo.user = user
-        mydevolo.password = password
-        if self.show_advanced_options:
-            mydevolo.url = user_input[CONF_MYDEVOLO]
-        else:
-            mydevolo.url = DEFAULT_MYDEVOLO
+        return await self._connect_mydevolo(user_input)
+
+    async def async_step_zeroconf(self, discovery_info: DiscoveryInfoType):
+        """Handle zeroconf discovery."""
+        # Check if it is a gateway
+        if discovery_info.get("properties", {}).get("MT") in SUPPORTED_MODEL_TYPES:
+            await self._async_handle_discovery_without_unique_id()
+            return await self.async_step_zeroconf_confirm()
+        return self.async_abort(reason="Not a devolo Home Control gateway.")
+
+    async def async_step_zeroconf_confirm(self, user_input=None):
+        """Handle a flow initiated by zeroconf."""
+        if user_input is None:
+            return self._show_form(step_id="zeroconf_confirm")
+        return await self._connect_mydevolo(user_input)
+
+    async def _connect_mydevolo(self, user_input):
+        """Connect to mydevolo."""
+        mydevolo = configure_mydevolo(conf=user_input)
```

Comment:
We're potentially moving directly to the user step form from here. We shouldn't do that. Each form is private to its step method. Await the step method to change step instead.